### PR TITLE
Prokop damp func

### LINKED ISSUE #48: Choose damping function for vdW
It was noted in #35 that the stand-alone vdW calculation has some unusual term, which was used as a quick solution for damping the force to prevent instability. There exist better functions for the damping, e.g. https://manual.q-chem.com/5.2/Ch5.S7.SS2.html. We should choose one that has a good compromise between stability and speed.

### PR DESCRIPTION
**Pull request by ProkopHapala**
fixes #48 

So I implemented 4 other damping function. In the process of testing I also corrected bug in PyridineDensOverlap example.

@NikoOinonen : When I plot the new damped vdW potential in 3D as isosurface in VESTA it still does not look like the atoms, just as a blob, but I think it is fine. Actually I think the old daping (with constant) was not that bad (considering how simple it was). 

In general the main question is how to choose the parameter ADamp (streght of damping). It ADamp is large the damping affect the potential at r=Rmin, if ADamp is small the potential is deep and it can cause the probe particle to fall inside atom, which is unrealistic. But in order to optimize ADamp properly it would requite e.g. to fit it with respect to DFT calculations (which is actually what Ruben Perez did).

### CODE REVIEW

File: cpp/Forces.h
Comment (NikoOinonen) [inline cpp/Forces.h:207]: I have usually seen the energy parameter with prefactor 4 instead of 2, like on wikipedia: https://en.wikipedia.org/wiki/Lennard-Jones_potential. Or maybe the parameter is defined somehow differently here?

Comment (ProkopHapala) [reply cpp/Forces.h:207]: It depends if you use `(sigma/r)^6` (sigma is position of node) or `(r0/r)^6` (r0 is position of minimum) 

see 
n-exp form in
https://en.wikipedia.org/wiki/Lennard-Jones_potential#Alternative_notations_of_the_Lennard-Jones_potential

File: cpp/Forces.h
Comment (NikoOinonen) [inline cpp/Forces.h:210]: I'm not convinced that this formula for the force is correct. From my calculation, the exponent for `u` in the numerator should be 5 instead of 4, and to make the unit vector for `dR` I think the denominator should be `sqrt(r2)`, so something like
```C
double fr = -E *e*( 6*u4*sqrt(u2) + dD*ADamp) / sqrt(r2);
```
Also, is the sign correct? The force is the negative derivative of the energy, but you also get one minus sign from the derivative of the denominator.

File: cpp/Forces.h
Comment (NikoOinonen) [inline cpp/Forces.h:154]: Do these commented-out blocks serve a purpose? I would just delete them to make it cleaner, if they are not required for later.

File: generateElFF.py
Comment (NikoOinonen) [inline generateElFF.py:65]: Remove the debug prints before merge.

File: cpp/ProbeParticle.cpp
Comment (NikoOinonen) [inline cpp/ProbeParticle.cpp:392]: Remove the debug print before merge.

File: cpp/Forces.h
Comment (NikoOinonen) [reply cpp/Forces.h:207]: Ok, just wanted to make sure it's consistent.

Comment (ProkopHapala) [reply cpp/Forces.h:207]: yes, I think it is good you check it. Because I often do mistakes of this king. So does it seems correct to you now?

File: cpp/ProbeParticle.cpp
Comment (ProkopHapala) [reply cpp/ProbeParticle.cpp]: OK, I removed the debug prints

File: cpp/Forces.h
Comment (NikoOinonen) [reply cpp/Forces.h]: I think that would be true without the damping function, but you are also multiplying the damping function by `r` which would cause some distortion to the damping of the force.

`invR2` here is the inverse of the square of the LJ parameter, isn't it, so it's not the same as `1 / sqrt(r2) = 1 / |dR|` which I thought of as the normalization for `dR` as you said.

For the minus signs, I think there should be three, one for the energy, one for the fact that `F = -dE/dr`, and one from the derivative of `1/f(r)`. But I think you only have two, because one is already in the energy, and the other two should cancel out in the force formula.

Comment (ProkopHapala) [reply cpp/Forces.h]: * Notice that  dD returned from Rfunc is also missing one power of `r` exactly for this reason that it is expected to be multiplied by `dR`. Therefore dD = (d Rfunc(r)/dr)/r` rater than just `d Rfunc(r)/dr`
* you are right invR2 si 1/R0^2 not 1/r^2, I overlooked that while replying you. But still the code is fine (I think):

```
E(r) = 1/( r^6 + Rfunc(r) )
F(r) =  dR *   -E(r)^2 * ( 6*r^4 + (dRfunc(r)/dr)/r )
```

* ad sign - I always make mistakes in this. Never shure if we store `F=-dE/dr` or `F=dE/dr`. But notice that the test here works (meaning the sign convecction is the same as for un-damped L-J and vdW )
https://github.com/Probe-Particle/ppafm/issues/48#issuecomment-1429919012

please try tu run the test here if you don't believe me:
https://github.com/Probe-Particle/ppafm/blob/prokop-damp-func/dev/test_vdwDaping.py

Comment (NikoOinonen) [reply cpp/Forces.h]: Ah, I didn't see that `dD` is missing a factor of `r`. Then it makes sense. I'm still a bit confused by the sign convention, but it looks like it comes out correct in the end, so I guess it's fine.

**Review by NikoOinonen — APPROVED** [review]
Looks good to me now.

**Comment by ProkopHapala** [comment]
OK, @NikoOinonen reviewed it quite troughfully. Therefore, should I wait for @ondrejkrejci  @yakutovicha or should I merge it straightaway ?

**Review by yakutovicha — APPROVED** [review]
No issues from my side, feel free to merge that

File: generateElFF.py
Comment (ondrejkrejci) [inline generateElFF.py:98]: What if tip is calcualated by aims or cp2k? Some more general loading would be better, at least in the future.

File: ppafm/defaults/params.ini
Comment (ondrejkrejci) [inline ppafm/defaults/params.ini:36]: Once more here it is writing me no free line, but I have a feeling that a free line would actually make problems.

**Review by ondrejkrejci — APPROVED** [review]
All the comments are minor, thus I approve. Still my comments to [examples/pyridineDensOverlap/run.sh-](https://github.com/Probe-Particle/ppafm/pull/87/files#diff-8f87b7d6c52495365919e44e7ee53aa3afcea7eae88c0fd45d7d7c29bc0b94c5) are worth of checking.

File: cpp/ProbeParticle.cpp
Comment (ProkopHapala) [reply cpp/ProbeParticle.cpp:393]: it is not used in this function, it is global variable used for walking over coef array in [evalCell()](https://github.com/Probe-Particle/ppafm/blob/4598d7d56f7cf04d3004714ed950e68d689c7889/cpp/ProbeParticle.cpp#L220)

it is not very readable and not-good practice, but it allows to use generic function to evaluate all interactions

File: dev/test_vdwDaping.py
Comment (ProkopHapala) [reply dev/test_vdwDaping.py:105]: I would keep that for formater

File: examples/pyridineDensOverlap/run.sh-
Comment (ProkopHapala) [reply examples/pyridineDensOverlap/run.sh-:16]: this is link to the original file which is missing so I replaced by new links

File: generateElFF.py
Comment (ProkopHapala) [reply generateElFF.py:98]: Yes, but keep this for future (I do not have time to test it now)

**Comment by yakutovicha** [comment]
@ProkopHapala, please squash-merge

**Pull request merged**
